src/transformStateWithClones.js

```diff
@@ -7,7 +7,26 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let newState = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      newState = { ...newState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      const updatedState = { ...newState };
+
+      for (const keyToRemove of action.keysToRemove) {
+        delete updatedState[keyToRemove];
+      }
+      newState = updatedState;
+    } else if (action.type === 'clear') {
+      newState = {};
+    }
+    states.push({ ...newState });
```

Comment:
Instead of cloning the 'newState' object inside the loop for each action, you can push the cloned state after the switch statement to adhere to the DRY (Don't Repeat Yourself) principle.